### ISSUE TITLE
Add function for random bool in math/rand

### ISSUE THREAD

**Issue by kirinrastogi** [golang/go#23804]
I want to be able to call `rand.Bool()` and receive a pseudo-random bool when I import `math/random`

**Comment by mdcnz** [comment]
[rand.Intn()](https://golang.org/pkg/math/rand/#Intn)

**Comment by kirinrastogi** [comment]
Yes I believe it is worth it. `rand.Intn(2) == 0` is not as elegant as `rand.Bool()`. This is more clear

**Comment by JackyChiu** [comment]
> How about math.IntN(2) == 0? Is it really worth adding a new function for this?

I'd argue that it's worth doing. It's not solving a hard problem but it's making it easier to read. Similar to how [sync.Waitgroup.Done()](https://golang.org/src/sync/waitgroup.go?s=3246:3273#L89) is an just an extra 1 line function but makes it more expressive to read for the programmer.

**Comment by ALTree** [comment]
I don't have an opinion of this, but I have a small argument in favour.

The way the `rand` package use the randomness source is quite intricate. A lot of the `rand` functions do not directly use the generator, but call other functions and then manipulate the input in some way to build a value of the requested type. For this reason, a lot of the "obvious" way to get a random bool results in very deep call stacks. 

Take for example the simple method Ian proposed above: `rand.Intn(2) == 0`. This is what happens:

We call r.Intn(2), which checks that the argument fits an int32,  and then call calls r.Int31n(2), which checks that the arguments is a power of two, and then calls r.Int31(), which calls r.Int63() and then shifts the result and cast it into an int32. 

This is what the call stack looks like:

```
  math/rand.(*Rand).Int63(0xc42007a1e0, 0xff)
	math/rand/rand.go:84 +0x22
  math/rand.(*Rand).Int31(0xc42007a1e0, 0xc42002def0)
	math/rand/rand.go:101 +0x2b
  math/rand.(*Rand).Int31n(0xc42007a1e0, 0x13b2c00000002, 0x390ffd3f)
	math/rand/rand.go:133 +0xb4
  math/rand.(*Rand).Intn(0xc42007a1e0, 0x2, 0x5a832bc1)
	math/rand/rand.go:174 +0x45
  math/rand.Intn(0x2, 0x4a91e1)
	math/rand/rand.go:331 +0x37 
```

An implementation that directly call the function that is the nearest to the generator, like this: `Int63() % 2 == 0`, on the other way, has a much smaller call stack, and it about 20% faster on my machine in a quick benchmark I wrote.

**Comment by bcmills** [comment]
> We call r.Intn(2), which checks that the argument fits an int32, and then call calls r.Int31n(2), which checks that the argument is a power of two, and then calls r.Int31(), which calls r.Int63() and then shifts the result and cast it into an int32.

All of those checks and conversions should be easy to inline and constant-fold away.
If they are not inlined and constant-folded, that is arguably a compiler bug, not an API bug.

**Comment by faiface** [comment]
If you're used to `rand.Intn(2) == 0`, *which you are* after you've used it 2-3x, it's as expressive as `rand.Bool()` would be.

**Comment by rsc** [comment]
Leaving this for math/rand v2 discussions. Please just use r.Intn(2) == 0 for now.

**Comment by CAFxX** [comment]
Whereas I agree that as defined this does not add much (except, maybe, avoid biases in known bad bits), I would suggest to consider adding a probability to that function:

```
// Bool returns the outcome of a Bernoulli random variable
// with the specified probability of returning true.
// Bool panics if p < 0 or p > 1.
func Bool(pTrue float64) bool
```

while this has an obvious implementation (`return Float64() < p`), this could benefit from constant propagation to completely avoid using floating points at runtime (e.g. convert the constant from the float64 range [0 1] to [0 MaxUint64] and then `return Uint64() < pUint64`)

**Comment by seankhliao** [comment]
8.2k hits for the current pattern of intn(2) ?? 0
https://github.com/search?q=language%3Ago+%2Frand.Int%5BnN%5D%5C%282%5C%29+%28%21%3D%7C%3D%3D%7C%3E%29+0%2F&type=code
though most call sites seem ok with just using that inline and not using a dedicated function for it
